crates/rethnet_evm/src/debug_trace.rs

```diff
@@ -0,0 +1,432 @@
+    }
+
+    Err(DebugTraceError::InvalidTransactionHash {
+        transaction_hash: *transaction_hash,
+        block_number: block_env.number,
+    })
+}
+
+/// Config options for `debug_trace_transaction`
+#[derive(Debug, Default, Clone)]
+pub struct DebugTraceConfig {
+    /// Disable storage trace.
+    pub disable_storage: bool,
+    /// Disable memory trace.
+    pub disable_memory: bool,
+    /// Disable stack trace.
+    pub disable_stack: bool,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum DebugTraceError<BlockchainErrorT, StateErrorT> {
+    /// Invalid hardfork spec argument.
+    #[error("Invalid spec id: {spec_id:?}. `debug_traceTransaction` is not supported prior to Spurious Dragon")]
+    InvalidSpecId { spec_id: SpecId },
+    /// Invalid transaction hash argument.
+    #[error("Transaction hash {transaction_hash} not found in block {block_number}")]
+    InvalidTransactionHash {
+        transaction_hash: B256,
+        block_number: U256,
+    },
+    #[error(transparent)]
+    SignatureError(#[from] SignatureError),
+    #[error(transparent)]
+    TransactionError(#[from] TransactionError<BlockchainErrorT, StateErrorT>),
+}
+
+/// Result of a `debug_traceTransaction` call.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct DebugTraceResult {
+    /// Whether transaction was executed successfully.
+    pub pass: bool,
+    /// All gas used by the transaction.
+    pub gas_used: u64,
+    /// Return values of the function.
+    pub output: Option<Bytes>,
+    /// The EIP-3155 debug logs.
+    pub logs: Vec<DebugTraceLogItem>,
+}
+
+/// The output of an EIP-3155 trace.
+/// The required fields match <https://eips.ethereum.org/EIPS/eip-3155#output> except for
+/// `returnData` and `refund` which are not used currently by Hardhat.
+/// The `opName`, `error`, `memory` and `storage` optional fields are supported as well.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub struct DebugTraceLogItem {
+    /// Program Counter
+    pub pc: u64,
+    /// Op code
+    pub op: u8,
+    /// Gas left before executing this operation as hex number.
+    pub gas: String,
+    /// Gas cost of this operation as hex number.
+    pub gas_cost: String,
+    /// Array of all values (hex numbers) on the stack
+    pub stack: Option<Vec<String>>,
+    /// Depth of the call stack
+    pub depth: u64,
+    /// Size of memory array.
+    pub mem_size: u64,
+    /// Name of the operation.
+    pub op_name: String,
+    /// Description of an error.
+    pub error: Option<String>,
+    /// Array of all allocated values as hex strings.
+    pub memory: Option<Vec<String>>,
+    /// Map of all stored values with keys and values encoded as hex strings.
+    pub storage: Option<HashMap<String, String>>,
+}
+
+// Based on https://github.com/bluealloy/revm/blob/70cf969a25a45e3bb4e503926297d61a90c7eec5/crates/revm/src/inspector/tracer_eip3155.rs
+// Original licensed under the MIT license.
+struct TracerEip3155 {
```

Comment:
Could you split the tracer into a separate file? We'll need to use this for `traceCall` too

Reply:
Sure, ok if I do that as part of https://github.com/NomicFoundation/edr/issues/192? Wanna think more about the best approach there.

Reply:
Sure